p2p/protocol/identify/peer_loop.go

```diff
@@ -0,0 +1,209 @@
+package identify
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	pb "github.com/libp2p/go-libp2p/p2p/protocol/identify/pb"
+
+	ggio "github.com/gogo/protobuf/io"
+)
+
+var errProtocolNotSupported = errors.New("protocol not supported")
+
+type peerHandler struct {
+	ids *IDService
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	pid peer.ID
+
+	msgMu         sync.RWMutex
+	idMsgSnapshot *pb.Identify
+
+	pushCh     chan struct{}
+	deltaCh    chan struct{}
+	evalTestCh chan func() // for testing
+}
+
+func newPeerHandler(pid peer.ID, ids *IDService, initState *pb.Identify) *peerHandler {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ph := &peerHandler{
+		ids:    ids,
+		ctx:    ctx,
+		cancel: cancel,
+		pid:    pid,
+
+		idMsgSnapshot: initState,
+
+		pushCh:     make(chan struct{}, 1),
+		deltaCh:    make(chan struct{}, 1),
+		evalTestCh: make(chan func()),
+	}
+
+	ph.wg.Add(1)
+	go ph.loop()
+	return ph
+}
+
+func (ph *peerHandler) close() error {
+	ph.cancel()
+	ph.wg.Wait()
+	return nil
+}
+
+// per peer loop for pushing updates
+func (ph *peerHandler) loop() {
+	defer ph.wg.Done()
+
+	for {
+		select {
+		// our listen addresses have changed, send an IDPush.
+		case <-ph.pushCh:
+			dp, err := ph.openStream([]string{IDPush})
+			if err != nil {
+				continue
+			}
+
+			mes := &pb.Identify{}
+			ph.ids.populateMessage(mes, dp.Conn())
+
+			ph.msgMu.Lock()
+			ph.idMsgSnapshot = mes
+			ph.msgMu.Unlock()
+
+			ph.sendMessage(dp, mes)
```

Comment:
errors silently dropped here and in delta below

Reply:
This has been fixed by logging them here.

Reply:
This has been fixed.